ppo/agents/advantage.py

```diff
@@ -0,0 +1,40 @@
+import jax.numpy as jnp
+from dataclasses import replace
+
+
+def general_advantage_estmation(agent, last_timestep, discount=0.99, gae_lambda=0.95):
+    """Estimate advantage function
+
+    Args:
+        agent (_type_): _description_
+        last_timestep (_type_): _description_
+        discount (float, optional): _description_. Defaults to 0.99.
+        gae_lambda (float, optional): _description_. Defaults to 0.95.
+
+    Returns:
+        _type_: _description_
+    """
+    trajectory = agent.replay_buffer._memory
+    last_value = agent.get_value(last_timestep.observation)
+    last_done = last_timestep.last()
+    advantages = jnp.zeros(len(trajectory))
+    lastgaelam = 0.0
+    for t in reversed(range(len(trajectory))):
+        if t == len(trajectory) - 1:
+            not_done_tp1 = 1.0 - last_done
+            value_tp1 = last_value
+        else:
+            not_done_tp1 = 1.0 - trajectory[t].done_tp1
+            value_tp1 = trajectory[t + 1].value_t
+        delta = trajectory[t].reward_tp1 + discount * value_tp1 * not_done_tp1 - trajectory[t].value_t
+        advantages = advantages.at[t].set((delta + discount * gae_lambda * not_done_tp1 * lastgaelam)[0])
+        lastgaelam = advantages[t]
+
+    return advantages
+
+
+def add_advantage(advantages, agent):
```

Comment:
It might be "cleaner" to add a method to `VanillaPPO`.
This method would call another method of `BaseReplayBuffer` adding the advantage to the memory.

This means that I would suggest adding two methods, one in `VanillaPPO` and one in `BaseReplayBuffer`.

In the main loop it would look like:
```Python
agent.add_advantage(advantage)
```